odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1516 @@
+        convex : `bool`, optional
+            If `True`, assume that the functional is convex
+        grad_lipschitz : 'float', optional
+            The Lipschitz constant of the gradient.
+        """
+
+        Operator.__init__(self, domain=domain,
+                          range=domain.field, linear=linear)
+        self._is_smooth = bool(smooth)
+        self._is_convex = bool(convex)
+        self._is_concave = bool(concave)
+        self._grad_lipschitz = float(grad_lipschitz)
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional.
+
+        Notes
+        -----
+        The operator that corresponds to the mapping
+
+        .. math::
+
+            x \\to \\nabla f(x)
+
+        where :math:`\\nabla f(x)` is the element used to evaluated
+        derivatives in a direction :math:`d` by
+        :math:`\\langle \\nabla f(x), d \\rangle`.
+        """
+        raise NotImplementedError
+
+    def proximal(self, sigma=1.0):
+        """Return the proximal operator of the functional.
+
+        Parameters
+        ----------
+        sigma : positive float, optional
+            Regularization parameter of the proximal operator.
+
+        Returns
+        -------
+        out : `Operator`
+            Domain and range equal to domain of functional.
+
+        Notes
+        -----
+        Note that ``Functional.proximal`` is a `Proximal factory`.
+        """
+        raise NotImplementedError
+
+    @property
+    def conjugate_functional(self):
+        """Convex conjugate functional of the functional.
+
+        Notes
+        -----
+        The convex conjugate functional of a convex functional :math:`f(x)`,
+        defined on a Hilber space, is defined as the functional
+
+        .. math::
+
+            f^*(x^*) = \\sup_{x} \{ \\langle x^*,x \\rangle - f(x)  \}.
+
+        See, e.g., [Lue1969]_, [Roc1970]_.
+        """
+        raise NotImplementedError
+
+    def derivative(self, point):
+        """Returns the derivative operator in the given point.
+
+        This function returns the linear operator
+
+            ``x --> <x, grad_f(point)>``,
+
+        where ``grad_f(point)`` is the gradient of the functional in the point
+        ``point``.
+
+        Parameters
+        ----------
+        point : `LinearSpaceVector`
+            The point in which the gradient is evaluated.
+
+        Returns
+        -------
+        out : `DerivativeOperator`
+            The linear operator that maps ``x --> <x, grad_f(point)>``.
+        """
+        functional = self
+
+        # TODO: Update doc for this operator/functional
+        # Also: it is a functional, not an operator, right?
+        class DerivativeOperator(Functional):
+            def __init__(self):
+                super().__init__(functional.domain, linear=True)
+
+            self.point = point
+
+            def _call(self, x):
+                return x.inner(functional.gradient(point))
+
+        return DerivativeOperator()
```

Comment:
Comment, short version:
This can be written in one line, `return self.gradient(point).T`.

Longer version:
The above code will return an `Operator`. If you want it to return a functional, you need to override the `T` property of `Functional`.